src/libopensc/card-esteid2025.c

```diff
@@ -0,0 +1,224 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <string.h>
+
+#include "asn1.h"
+#include "gp.h"
+#include "internal.h"
+
+#define SIGNATURE_PAYLOAD_SIZE 0x30
+
+static const struct sc_atr_table esteid_atrs[] = {
+		{"3b:ff:96:00:00:80:31:fe:43:80:31:b8:53:65:49:44:64:b0:85:05:10:12:23:3f:1d", NULL, "EstEID 2025", SC_CARD_TYPE_ESTEID_2025, 0, NULL},
+		{NULL,									 NULL, NULL,	   0,			      0, NULL}
+};
+
+static const struct sc_aid THALES_AID = {
+		{0xA0, 0x00, 0x00, 0x00, 0x63, 0x50, 0x4B, 0x43, 0x53, 0x2D, 0x31, 0x35},
+		12
+};
+
+static const struct sc_card_operations *iso_ops = NULL;
+static struct sc_card_operations esteid_ops;
+
+static struct sc_card_driver esteid2025_driver = {"EstEID 2025", "esteid2025", &esteid_ops, NULL, 0, NULL};
+
+#define SC_TRANSMIT_TEST_RET(card, apdu, text) \
+	{ \
+		LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed"); \
+		LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), text); \
+	}
+
+static int
+esteid_match_card(sc_card_t *card)
+{
+	int i = _sc_match_atr(card, esteid_atrs, &card->type);
+
+	if (i >= 0 && gp_select_aid(card, &THALES_AID) == SC_SUCCESS) {
+		card->name = esteid_atrs[i].name;
+		return 1;
+	}
+	return 0;
+}
+
+static int
+esteid_select_file(struct sc_card *card, const struct sc_path *in_path, struct sc_file **file_out)
+{
+	u8 resp[SC_MAX_APDU_RESP_SIZE];
+	size_t resplen = sizeof(resp);
+	int r;
+	struct sc_file *file = NULL;
+	struct sc_apdu apdu;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	// Only support full paths
+	if (in_path->type != SC_PATH_TYPE_PATH) {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	}
+
+	sc_format_apdu_ex(&apdu, card->cla, 0xA4, 0x08, 0x04, in_path->value, in_path->len, resp, resplen);
+	SC_TRANSMIT_TEST_RET(card, apdu, "SELECT failed");
+	if (file_out != NULL) {
+		file = sc_file_new();
+		if (file == NULL)
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+		r = iso_ops->process_fci(card, file, resp, resplen);
+		LOG_TEST_RET(card->ctx, r, "Process fci failed");
+		*file_out = file;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+esteid_set_security_env(sc_card_t *card, const sc_security_env_t *env, int se_num)
+{
+	struct sc_apdu apdu;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (env == NULL || env->key_ref_len != 1)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	sc_log(card->ctx, "algo: %lu operation: %d keyref: %d", env->algorithm, env->operation, env->key_ref[0]);
+
+	if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN) {
+		const u8 cse_crt_sig[] = {0x80, 0x01, 0x54, 0x84, 0x01, env->key_ref[0]};
```

Comment:
Is it guaranteed that cse_crt_sig and cse_crt_der will still be present for the transmit when going out of scope when declared as const? I'm not sure about the actual c standard, but I think it would be more intuitive to put those definitions at the top of the function.